okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -266,7 +237,63 @@ class ExchangeFinder(
     }
 
     eventListener.connectionAcquired(call, result)
-    return result
+    return ReusePlan(result)
+  }
+
+  internal sealed class Plan
+
+  /** Reuse an existing connection. */
+  internal class ReusePlan(
+    val connection: RealConnection,
+  ) : Plan()
+
+  /** Establish a new connection. */
+  internal inner class ConnectPlan(
+    route: Route,
+    val routes: List<Route>? = null,
+  ) : Plan() {
+    val connection = RealConnection(client.taskRunner, connectionPool, route)
+    var complete = false
+    var failure: IOException? = null
+
+    fun connect() {
```

Comment:
This is all working towards making it so we can run this function in a background thread.

Reply:
So for Happy Eyeballs, do you intend to have a HappyEyeballsConnectPlan? or multiple ConnectPlans for each possible route? Where do implementations of things like delays for additional results come in, or preferring routes?

Reply:
My best idea is we have a single thread, the call thread, act as coordinator.

Then all calls to Socket.connect() need to be on a TaskRunner.

The greatest challenge is that the coordinator needs to also do the DNS lookups, at least unless I extract that out.

Reply:
Because of sync call or just code structure? We have a lot of better options for DNS. Android DnsResolver, OkHttp DoH, other JVM options. Dns.SYSTEM would be a waste for Happy Eyeballs.

Reply:
Yeah, in this case it's just the concurrency that's tricky. We have a bunch of state on which routes we need to try and it's currently pretty gnarly.